xcube/core/store/typeid.py

```diff
@@ -0,0 +1,124 @@
+    A type id denotes a type of data. It is used to group similar types of data and discern different types of data.
+    It can be used by stores to state what types of data can be read from and/or written to them.
+
+    A type id consists of a name and an arbitrary number of flags.
+    Flags can be used to further refine a type id if needed.
+    """
+
+    def __init__(self, name: str, flags: Set[str] = None):
+        assert_given(name, 'name')
+        self._name = name
+        self._flags = flags if flags is not None else set()
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def flags(self) -> Set[str]:
+        return self._flags
+
+    def __str__(self) -> str:
+        if len(self.flags) == 0:
+            return self.name
+        flag_part = ','.join(sorted(self.flags))
+        return f'{self.name}[{flag_part}]'
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    def __eq__(self, other) -> bool:
+        try:
+            other_type = self.normalize(other)
+        except TypeError:
+            return False
+        if self.name != other_type.name:
+            return False
+        return len(self.flags.symmetric_difference(other_type.flags)) == 0
+
+    def is_compatible(self, other: Union[str, "TypeId"]) -> bool:
+        """
+        Tests whether another type id is compatible with this type id.
+        This is a weaker relationship than *equals()*.
+        Two type ids are considered compatible when they have the same name and the other type id has all the flags
+        that this type id has. The other type id may also have additional flags.
+
+        :param other: Another type id, as string or *TypeId*.
+        :return: Whether the other type id is compatible with this type id
+        """
+        other_type = self.normalize(other)
+        if self.name != '*' and self.name != other_type.name:
+            return False
+        return len(self.flags.difference(other_type.flags)) == 0
+
+    def __hash__(self) -> int:
+        return hash(self.name) + 16 * hash(frozenset(self.flags))
+
+    @classmethod
+    def normalize(cls, type_id: Union[str, "TypeId"]) -> "TypeId":
+        if isinstance(type_id, TypeId):
+            return type_id
+        if isinstance(type_id, str):
+            return cls.parse(type_id)
+        raise TypeError('type_id must be of type "str" or "TypeId"')
+
+    @classmethod
+    def parse(cls, type_id: str) -> "TypeId":
+        if '[' not in type_id:
+            return TypeId(type_id)
+        if not type_id.endswith(']'):
+            raise SyntaxError(f'"{type_id}" cannot be parsed: No end brackets found')
+        name = type_id.split('[')[0]
+        flags = type_id.split('[')[1].split(']')[0].split(',')
+        return TypeId(name, flags=set(flags))
+
+TYPE_ID_ANY = TypeId('*')
+TYPE_ID_DATASET = TypeId('dataset')
+TYPE_ID_CUBE = TypeId('dataset', flags={'cube'})
+TYPE_ID_MULTI_LEVEL_DATASET = TypeId('mldataset')
```

Comment:
Actually, this is the correct definition:

```suggestion
TYPE_ID_MULTI_LEVEL_DATASET = TypeId('dataset', flags={'multilevel'})
```

However, a `MultiLevelDataset` is not yet an instance  of `xarray.Dataset` but that should be possible later.